pythia/modules/decoders.py

```diff
@@ -47,3 +49,78 @@ def forward(self, encoder_output, batch):
         scores = torch.sum(answer_options * encoder_output, dim=2)
 
         return scores
+
+
+class CaptioningDecoder(nn.Module):
+    def __init__(
+        self,
+        attention_dim,
+        embed_dim,
+        decoder_dim,
+        vocab_size,
+        features_dim=2048,
+        dropout=0.5,
+    ):
+        super(CaptioningDecoder, self).__init__()
+
+        self.features_dim = features_dim
+        self.attention_dim = attention_dim
+        self.embed_dim = embed_dim
+        self.decoder_dim = decoder_dim
+        self.vocab_size = vocab_size
+        self.dropout = dropout
+
+        self.attention = CaptionAttention(
+            features_dim, decoder_dim, attention_dim
+        )
+
+        self.dropout = nn.Dropout(p=self.dropout)
+        self.top_down_attention = nn.LSTMCell(
+            embed_dim + features_dim + decoder_dim, decoder_dim, bias=True
+        )
+        self.language_model = nn.LSTMCell(
+            features_dim + decoder_dim, decoder_dim, bias=True
+        )
+        self.fc = nn.utils.weight_norm(
+            nn.Linear(decoder_dim, vocab_size)
+        )
+        self.init_weights()
+     
+    def init_weights(self):
+        self.fc.bias.data.fill_(0)
+        self.fc.weight.data.uniform_(-0.1, 0.1)
+
+    def init_hidden_state(self, batch_size):
+        h = torch.zeros(batch_size, self.decoder_dim).to("cuda")
```

Comment:
As mentioned above, please use `new` creation ops or use this and move to proper device.